src/types/timezone.jl

```diff
@@ -1,4 +1,30 @@
-const TIME_ZONE_CACHE = Dict{String,Tuple{TimeZone,Class}}()
+# ---------- Time Zone Cache setup ----------------------------------------------------------------
+# Thread-local timezone caches to support multithreaded tasks simultaneously constructing TimeZones.
+# See https://github.com/JuliaTime/TimeZones.jl/issues/342 for more information.
+
+const THREADLOCAL_TIME_ZONE_CACHES = Dict{String,Tuple{TimeZone,Class}}[]
+
+@inline default_tz_cache() = default_tz_cache(Threads.threadid())
+@noinline function default_tz_cache(tid::Int)
+    0 < tid <= length(THREADLOCAL_TIME_ZONE_CACHES) || _tz_cache_length_assert()
+    if @inbounds isassigned(THREADLOCAL_TIME_ZONE_CACHES, tid)
+        @inbounds TZ_CACHE = THREADLOCAL_TIME_ZONE_CACHES[tid]
+    else
+        TZ_CACHE = eltype(THREADLOCAL_TIME_ZONE_CACHES)()
+        @inbounds THREADLOCAL_TIME_ZONE_CACHES[tid] = TZ_CACHE
+    end
+    return TZ_CACHE
+end
+@noinline _tz_cache_length_assert() =  @assert false "0 < tid <= length(THREADLOCAL_TIME_ZONE_CACHES)"
+
+module __InitThreadLocalCaches  # This module is only here to house the __init__() function.
+using ..TimeZones: THREADLOCAL_TIME_ZONE_CACHES
+function __init__()
+    resize!(empty!(THREADLOCAL_TIME_ZONE_CACHES), Threads.nthreads()) # ensures that we didn't save a bad object
+end
+end
```

Comment:
This pattern is a bit unconventional, and i don't know if there are any negative consequences to it?

There's already an `__init__()` for the module, but it seems annoying to put the cache initialization there. So i created a new module just to house the `__init__` closer to the definition.

Definitely happy to move this - lemme know what you prefer :)

Reply:
Probably a cleaner compromise would be to call a `_tz_cache_init` function defined here from root `__init__`

Reply:
agreed - i'll change that.